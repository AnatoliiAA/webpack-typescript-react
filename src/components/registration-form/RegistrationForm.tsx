// import React from "react";
import { useForm, SubmitHandler, Control } from "react-hook-form";
import { Input } from "../form-inputs/Input/Input";
import { SelectInput } from "../form-inputs/Select/Select";
import { Checkbox } from "../form-inputs/Checkbox/Checkbox";
import {
  Form,
  FormButtons,
  SubmitButton,
  BackButton,
} from "./RegistrationForm.css";
import inputData from "./data.json";

interface IFormInput {
  storeName: string;
  industry: { label: string; value: string };
  isRegistered: boolean;
  legalName: string;
  vatNumber: string;
  address: string;
  zipCode: string;
  city: string;
  country: { label: string; value: string };
}

const createInput = (
  type: string,
  name: string,
  attrs: {
    id?: string;
    labelText?: string;
    placeholder?: string;
    control?: Control<IFormInput, object>;
    options?: Array<{ value: string; label: string }>;
    testId: string;
    errors?: object;
    rules?: object;
    width?: string;
  }
): JSX.Element =>
  type === "text" ? (
    <Input type={type} name={name} {...attrs} key={attrs.id} />
  ) : type === "select" ? (
    <SelectInput name={name} {...attrs} key={attrs.id} />
  ) : type === "checkbox" ? (
    <Checkbox name={name} {...attrs} key={attrs.id} />
  ) : (
    <p>Invalid type definition</p>
  );

export const RegistrationForm = (): JSX.Element => {
  const {
    control,
    handleSubmit,
    formState: { errors },
  } = useForm<IFormInput>({ mode: "onChange" });

  const onSubmit: SubmitHandler<IFormInput> = async (data) => {
    const response = await fetch("https://www.google.com", {
      method: "POST",
      mode: "no-cors",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      {inputData.inputs.map(
        ({
          name,
          type,
          id,
          placeholder,
          testId,
          options,
          labelText,
          rules,
          width,
        }) =>
          createInput(type, name, {
            id,
            placeholder,
            testId,
            options,
            labelText,
            control,
            rules,
            errors,
            width,
          })
      )}
      <FormButtons>
        <BackButton>Go Back</BackButton>
        <SubmitButton type="submit" data-testid="submit-button">
          Save and continue
        </SubmitButton>
      </FormButtons>
    </Form>
  );
};
