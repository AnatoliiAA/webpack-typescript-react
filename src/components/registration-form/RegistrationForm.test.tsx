import React from "react";
import {
  render,
  fireEvent,
  getByTestId,
  queryAllByText,
  waitFor,
  getByText,
  queryAllByRole,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import { RegistrationForm } from "./RegistrationForm";

const testIds = {
  submit: "submit-button",
  storeName: "store-name",
  industry: "industry",
  registered: "is-registered",
  legalName: "legal-name",
  vatNumber: "vat-number",
  address: "address",
  zipCode: "zip-code",
  city: "city",
  country: "country",
};

describe("Registration Form Tests", () => {
  const {
    submit,
    storeName,
    industry,
    registered,
    legalName,
    vatNumber,
    address,
    zipCode,
    city,
    country,
  } = testIds;

  describe("snapshot test", () => {
    it("should match snapshot", () => {
      const tree = renderer.create(<RegistrationForm />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("actions tests", () => {
    it("should show 7 errors on submitting an empty form", async () => {
      const { container } = render(<RegistrationForm />);
      const submitButton = getByTestId(container, submit);
      await waitFor(() => {
        fireEvent.click(submitButton);
      });
      const errors = queryAllByText(container, "This field is required.");
      expect(errors.length).toBe(7);
    });

    it("should show error if name is too short", async () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = getByTestId(container, storeName);
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "1" } });
      });
      const inputError = getByText(
        container,
        "Name should contain 2 characters or more"
      );
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on clearing the 'name' field", async () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = getByTestId(container, storeName);
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "1" } });
      });
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "" } });
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on unchecking the 'is registered' checkbox", async () => {
      const { container } = render(<RegistrationForm />);
      const registeredCheckbox = getByTestId(container, registered);
      await waitFor(() => {
        fireEvent.click(registeredCheckbox);
        fireEvent.click(registeredCheckbox);
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should show error if legal name is too short", async () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = getByTestId(container, legalName);
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "1" } });
      });
      const inputError = getByText(
        container,
        "Legal Name should contain 2 characters or more"
      );
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on clearing the 'legal name' field", async () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = getByTestId(container, legalName);
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "1" } });
      });
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "" } });
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should show error if address is too short", async () => {
      const { container } = render(<RegistrationForm />);
      const addressInput = getByTestId(container, address);
      await waitFor(() => {
        fireEvent.change(addressInput, { target: { value: "1" } });
      });
      const inputError = getByText(
        container,
        "Address should contain 2 characters or more"
      );
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on clearing the 'address' field", async () => {
      const { container } = render(<RegistrationForm />);
      const addressInput = getByTestId(container, address);
      await waitFor(() => {
        fireEvent.change(addressInput, { target: { value: "1" } });
      });
      await waitFor(() => {
        fireEvent.change(addressInput, { target: { value: "" } });
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on clearing the 'zip code' field", async () => {
      const { container } = render(<RegistrationForm />);
      const zipCodeInput = getByTestId(container, zipCode);
      await waitFor(() => {
        fireEvent.change(zipCodeInput, { target: { value: "1" } });
      });
      await waitFor(() => {
        fireEvent.change(zipCodeInput, { target: { value: "" } });
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should show error if city name is too short", async () => {
      const { container } = render(<RegistrationForm />);
      const cityInput = getByTestId(container, city);
      await waitFor(() => {
        fireEvent.change(cityInput, { target: { value: "1" } });
      });
      const inputError = getByText(
        container,
        "City name should contain 2 characters or more"
      );
      expect(inputError).toBeInTheDocument();
    });

    it("should show the error on clearing the 'city' field", async () => {
      const { container } = render(<RegistrationForm />);
      const cityInput = getByTestId(container, city);
      await waitFor(() => {
        fireEvent.change(cityInput, { target: { value: "1" } });
      });
      await waitFor(() => {
        fireEvent.change(cityInput, { target: { value: "" } });
      });
      const inputError = getByText(container, "This field is required.");
      expect(inputError).toBeInTheDocument();
    });

    it("should call onSubmit function if form was filled correctly", async () => {
      global.fetch = jest.fn(() => Promise.resolve()) as jest.Mock;
      const { container } = render(<RegistrationForm />);
      const submitButton = getByTestId(container, submit);
      const nameInput = getByTestId(container, storeName);
      const industrySelect = getByTestId(container, industry);
      const registeredCheckbox = getByTestId(container, registered);
      const legalNameInput = getByTestId(container, legalName);
      const vatNumberInput = getByTestId(container, vatNumber);
      const addressInput = getByTestId(container, address);
      const zipCodeInput = getByTestId(container, zipCode);
      const cityInput = getByTestId(container, city);
      const countrySelect = getByTestId(container, country);
      await waitFor(() => {
        fireEvent.change(nameInput, { target: { value: "Name" } });
        fireEvent.click(registeredCheckbox);
        fireEvent.change(industrySelect, { target: { value: "oil" } });
        fireEvent.change(cityInput, { target: { value: "City" } });
        fireEvent.change(legalNameInput, { target: { value: "Legal Name" } });
        fireEvent.change(vatNumberInput, { target: { value: "123" } });
        fireEvent.change(addressInput, { target: { value: "Address" } });
        fireEvent.change(zipCodeInput, { target: { value: "123" } });
        fireEvent.change(countrySelect, { target: { value: "ukraine" } });
        fireEvent.click(submitButton);
      });
      let errors = container.querySelectorAll("p");
      expect(errors.length).toBe(0);
    });
  });
});
