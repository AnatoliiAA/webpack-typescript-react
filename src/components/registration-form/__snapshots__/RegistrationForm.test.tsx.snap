// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration Form Tests snapshot test should match snapshot 1`] = `
<form
  className="css-1h619rn"
  onSubmit={[Function]}
>
  <div
    className="css-1qci0n7"
  >
    <label
      className="css-teqmh5"
      htmlFor="store-name"
    />
    <input
      className="css-bsvvfw"
      data-testid="store-name"
      id="store-name"
      name="storeName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Store Name"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-1qci0n7"
  >
    <label
      htmlFor="industry"
    />
    <select
      className="css-10l0zdv"
      data-testid="industry"
      id="industry"
      name="industry"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Industry"
      value=""
    >
      <option
        disabled={true}
        selected={true}
        value=""
      >
        Industry
      </option>
      <option
        value="health"
      >
        Health
      </option>
      <option
        value="oil"
      >
        Oil
      </option>
      <option
        value="ai"
      >
        Artificial Intelligence
      </option>
    </select>
  </div>
  <div
    className="css-1qci0n7"
  >
    <input
      className="css-wcq7w9"
      data-testid="is-registered"
      id="is-registered"
      name="isRegistered"
      onBlur={[Function]}
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="css-1qwhpv3"
      htmlFor="is-registered"
    >
      This store is registered business
    </label>
  </div>
  <div
    className="css-1qci0n7"
  >
    <label
      className="css-teqmh5"
      htmlFor="legal-name"
    />
    <input
      className="css-bsvvfw"
      data-testid="legal-name"
      id="legal-name"
      name="legalName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Company Legal Name"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-1qci0n7"
  >
    <label
      className="css-teqmh5"
      htmlFor="vat-number"
    />
    <input
      className="css-bsvvfw"
      data-testid="vat-number"
      id="vat-number"
      name="vatNumber"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="VAT EU Number"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-1qci0n7"
  >
    <label
      className="css-teqmh5"
      htmlFor="address"
    />
    <input
      className="css-bsvvfw"
      data-testid="address"
      id="address"
      name="address"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Address"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-1666a7r"
    width="30%"
  >
    <label
      className="css-teqmh5"
      htmlFor="zip-code"
    />
    <input
      className="css-bsvvfw"
      data-testid="zip-code"
      id="zip-code"
      name="zipCode"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="ZIP Code"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-ck8uiq"
    width="60%"
  >
    <label
      className="css-teqmh5"
      htmlFor="city"
    />
    <input
      className="css-bsvvfw"
      data-testid="city"
      id="city"
      name="city"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="City"
      type="text"
      value=""
    />
  </div>
  <div
    className="css-1qci0n7"
  >
    <label
      htmlFor="country"
    />
    <select
      className="css-10l0zdv"
      data-testid="country"
      id="country"
      name="country"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Country"
      value=""
    >
      <option
        disabled={true}
        selected={true}
        value=""
      >
        Country
      </option>
      <option
        value="ukraine"
      >
        Ukraine
      </option>
      <option
        value="poland"
      >
        Poland
      </option>
      <option
        value="slovakia"
      >
        Slovakia
      </option>
    </select>
  </div>
  <div
    className="css-1fzj550"
  >
    <a
      className="css-v28um5"
    >
      Go Back
    </a>
    <button
      className="css-5qt4gn"
      data-testid="submit-button"
      type="submit"
    >
      Save and continue
    </button>
  </div>
</form>
`;
